### DISCUSSION TITLE
.ebextentions powershell script run command

### DISCUSSION THREAD

**Discussion by MohamedSalahFCI** [PowerShell/PowerShell#19365] [Q&A]
￼￼

I have an aws pipeline and deploy stage will be execute on elastic beanstalk and i have a powershell script uploaded to s3 i want to run this script on ec2 with administrator prevelage .. i found .ebextentions to do this and first command to download script from s3 run perfect but second command that run script with administrator user was incorrect for many times 
My configuration file like : 
commands:
  01_download_script:
    command: powershell.exe -Command "Read-S3Object -BucketName <S3-Bucket-Name> -Key <S3-Object-Key> -File <Local-File-Path>"
  02_run_script:
    command: powershell.exe -Command "Start-Process powershell.exe -ArgumentList '-ExecutionPolicy Bypass -File <Local-File-Path>' -Verb RunAs"

**Comment by kasini3000** [comment]
try:
powershell.exe -Command "Start-Process powershell.exe -ArgumentList '-ExecutionPolicy Bypass -File d:\path.ps1 -Verb RunAs  '    "

**Reply by MohamedSalahFCI** [reply]
If this command will prompt the user to confirm the elevation of privileges before running the script ?

**Reply by kasini3000** [reply]
No one can bypass microsoft UAC

**Reply by MohamedSalahFCI** [reply]
Do u know anything about deploy.bat file in .ebextentions folder ?

**Reply by rhubarb-geek-nz** [reply]
> Do u know anything about deploy.bat file in .ebextentions folder ?

https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/customize-containers-windows-ec2.html

An Elastic Beanstalk instance is an EC2 instance, this page was linked from the Beanstalk pages.

I have used similar mechanisms on Linux based beanstalks where we deployed custom RPM at the instance launch time. This documentation refers to a similar mechanism for Windows where you provide a configuration containing a list of URL for MSI to install.

This would allow you to cleanly package your code as an MSI, (eg using [WiX](https://wixtoolset.org) ) and AWS would perform the MSI installation.

**Reply by MohamedSalahFCI** [reply]
```
container_commands:
  00-myscript:
    command: powershell.exe -ExecutionPolicy Bypass -Command "Start-Process powershell.exe -Verb runas -ArgumentList '-ExecutionPolicy Bypass','-NoProfile','-NonInteractive','-Command','.\\.ebextensions\\myscript.ps1','>', 'C:\\myOutput6.txt', '2>&1'"

``` in my update i set my script in tha path of .ebextensions and i want to run it with container command .. do u have any suggest ?

**Comment by rhubarb-geek-nz** [comment]
> I have an aws pipeline and deploy stage will be execute on elastic beanstalk and i have a powershell script uploaded to s3 i want to run this script on ec2 with administrator prevelage

What operating system are you using for your Elastic Beanstalk? You don't mention it, except you say "powershell.exe"; then this is the WindowsPowerShell, not the PowerShell Core.

The best way for running background tasks on a Windows machine is either as a Windows service or as a Scheduled Tasks.

What are you doing that requires administrator rights on a Beanstalk ? These should be stateless machines capable of being load balanced. There should be nothing of value on the machines themselves.

**Reply by MohamedSalahFCI** [reply]
I am using a single instance beanstalk type .. so my script installs a certificate and bind it to iis

**Reply by rhubarb-geek-nz** [reply]
> I am using a single instance beanstalk type .. so my script installs a certificate and bind it to iis

This must be a very common situation if using IIS and no load balancer, is there no provided mechanism? The standard way with AWS is to have the SSL certificate in the load balancer. 

Is this a purely private instance on a private subnet with no public access? If so is there any need for SSL?

**Reply by kasini3000** [reply]
Would you like to know about my project: it is like ansible,but not use YAML
Open source, free, cross platform(FOR WIN,AND LINUX), cross cloud, mass deployment of "Kasini3000" by devops powershell
https://gitee.com/chuanjiao10/kasini3000
tech:
https://www.cnblogs.com/piapia/p/16534765.html

no need upload ps1 script to s3,
from 'master' to 'node': remote call powershell ,iis module on win-server(or win client) by winrm .

** Any topic of my project should be submitted here as much as possible: **
https://github.com/kasini3000/kasini3000/issues

**Reply by rhubarb-geek-nz** [reply]
> Would you like to know about my project

How does this solve the problem of installing an SSL certificate into an IIS server running as a beanstalk/EC2 where the goal is to configure the SSL certificate at the point of instance creation?

**Reply by kasini3000** [reply]
```powershell
#on win,linux master ---> win,or linux node
# all powershell for win node.devops no need beanstalk

./k_run_ip -ip $win_server1_ip {

	#Get-Command -module WebAdministration
	#get-command -module IISAdministration
	#https://learn.microsoft.com/en-us/powershell/module/iisadministration/clear-iiscentralcertprovider?view=windowsserver2022-ps
	# command likes:

	#new site
	New-Item iis:\Sites\ -bindings @{protocol = "http";bindingInformation = ":80:domain.com" } -physicalPath D:\web\


	$cert = Get-Item cert:\LocalMachine\My\certname
	#or
	#$cert = Get-ChildItem cert:\LocalMachine\My | where-object { $_.Subject -like "*$hostname*" }
	#$cert.Thumbprint

	$binding = Get-WebBinding -Name $site -Protocol "https"

	# banding cert
	$binding.AddSslCertificate($cert.GetCertHashString(), "my")

}
```

**Comment by rhubarb-geek-nz** [comment]
Are you using a tool such as [PowerShell Snap-in: Configuring SSL with the IIS PowerShell Snap-in](https://learn.microsoft.com/en-us/iis/manage/powershell/powershell-snap-in-configuring-ssl-with-the-iis-powershell-snap-in)?

Is the problem that the rights of the current user don't include what is needed to run this script?

If the script is already running as an administrator can it give itself the required rights?

**Comment by rhubarb-geek-nz** [comment]
As @kasini3000 points out, even the mighty PowerShell cannot overcome Windows RunAs limitations. We have to work with what the given execution environment can do.

Although this is from 2016, does this still work? https://www.dagint.com/iis-aws-elastic-beanstalk-end-to-end-ssl-encryption/

I don't think this is a PowerShell issue, more of an IIS/AWS/Beanstalk issue, suggest follow up on AWS forums.

**Reply by kasini3000** [reply]
> even the mighty PowerShell cannot overcome Windows RunAs limitations

There are no RunAs limitations on winrm remoting

### powershell & Winrm's real advantages:
* All programs are at the thread level. You can transfer objects.
* Automatic serialization, deserialization, and remote delivery of objects.
* It is easy to pass values to the PS1 script,devops by ps1 script better than YAML.
* Alarm by win sound ;txt2vocie; desktop wallpaper; win termal background image.
* Enter. net dll files on https://nuget.org , Powershell is easy to use database clients,Easily store values into a database.

----

on you post:

* no need upload script to S3.
* 00-commands.config: liks my post by winrm
* https-instance-securitygroup.config: open port from aws
You can use aws cli, but it is recommended to use the aws powershell module:
```powershell
https://docs.aws.amazon.com/powershell/latest/reference/items/Grant-EC2SecurityGroupIngress.html
$ip1 = @{ IpProtocol="tcp"; FromPort="22"; ToPort="22"; IpRanges="203.0.113.25/32" }
$ip2 = @{ IpProtocol="tcp"; FromPort="3389"; ToPort="3389"; IpRanges="203.0.113.25/32" }
Grant-EC2SecurityGroupIngress -GroupName "my-security-group" -IpPermission @( $ip1, $ip2 )
```
* https-lb-passthrough.config:

```powershell
# register
Register-ELB2Target -TargetGroupArn 'arn:aws:elasticloadbalancing:us-east-1:123456789012:targetgroup/test-tg/a4e04b3688be1970' -Target @{Port = 80; Id = 'i-0672a4c4cdeae3111'}

# unregister
$targetDescription = New-Object Amazon.ElasticLoadBalancingV2.Model.TargetDescription
$targetDescription.Id = 'i-0672a4c4cdeae3111'
Unregister-ELB2Target -Target $targetDescription -TargetGroupArn 'arn:aws:elasticloadbalancing:us-east-1:123456789012:targetgroup/test-tg/a4e04b3688be1970'
#https://docs.aws.amazon.com/powershell/latest/reference/items/ElasticLoadBalancingV2_cmdlets.html
```
# So, powershell is bestest for devops & Batch management of cloud servers,thanks for powershell group

Powershell is the best engine, and Kasini3000 is a car made by this engine.

My tip,call powershell on jenkins:
https://gitee.com/chuanjiao10/kasini3000/wikis/%E5%8D%A1%E6%AD%BB%E4%BD%A03000%E7%9A%84jenkins%E4%BE%8B%E5%AD%90

**Reply by rhubarb-geek-nz** [reply]
> There are no RunAs limitations on winrm remoting

There is no remote command execution involved. The ebextensions are run by the server itself when it is created.
